File: 2021/cpp/day_9.cpp
//
// Created by trox667 on 09.12.21.
//

#include "day_9.hpp"
#include <set>

namespace day9 {
    [[nodiscard]] std::vector<int>
    heightmap_from_file(const std::vector<std::string> &lines) {
        auto heightmap = std::vector<int>{};
        for (const auto &line: lines) {
            for (const auto &c: line) {
                heightmap.emplace_back(c - '0');
            }
        }
        return heightmap;
    }

    int index(int x, int y, int width) {
        return y * width + x;
    }

    std::pair<int, int> position(int idx, int width) {
        return std::make_pair(idx % width, idx / width);
    }

    [[nodiscard]] std::vector<int>
    get_neighbors(int x, int y, int width, int height) {
        auto n = std::vector<int>{};
        if (x - 1 >= 0) n.emplace_back(index(x - 1, y, width));
        if (x + 1 < width) n.emplace_back(index(x + 1, y, width));
        if (y - 1 >= 0) n.emplace_back(index(x, y - 1, width));
        if (y + 1 < height) n.emplace_back(index(x, y + 1, width));
        return n;
    }

    [[nodiscard]] std::set<int>
    get_low_points(const std::vector<int> &heightmap, const int width,
                   const int height) {
        auto low_points = std::set<int>();
        for (auto i = 0; i < heightmap.size(); ++i) {
            auto pos = position(i, width);
            auto x = std::get<0>(pos);
            auto y = std::get<1>(pos);
            auto curr_height = heightmap[i];
            if (curr_height == 9) continue;

            auto neighbors = get_neighbors(x, y, width, height);
            auto values = std::vector<int>(neighbors.size());
            std::transform(neighbors.begin(), neighbors.end(), values.begin(),
                           [heightmap](auto a) { return heightmap[a]; });
            auto min = std::min_element(values.begin(), values.end());
            if (curr_height < *min) low_points.insert(i);
        }
        return low_points;
    }

    void
    get_basin(const int idx, const std::vector<int> &heightmap, const int width,
              const int height, std::set<int> &basin, std::set<int> &visited) {
        visited.insert(idx);
        auto pos = position(idx, width);
        auto x = std::get<0>(pos);
        auto y = std::get<1>(pos);
        auto curr_height = heightmap[idx];
        if (curr_height == 9) return;

        basin.insert(idx);

        auto neighbors = get_neighbors(x, y, width, height);
        for (const auto &n: neighbors) {
            if (!visited.contains(n) && heightmap[n] != 9) {
                get_basin(n, heightmap, width, height, basin, visited);
            }
        }
    }

    [[nodiscard]] long part1(const std::vector<std::string> &lines) {
        auto width = lines[0].size();
        auto height = lines.size();
        const auto &heightmap = heightmap_from_file(lines);
        const auto &low_points = get_low_points(heightmap, width, height);
        auto sum = 0;
        for (const auto &low_point: low_points) {
            sum += heightmap[low_point] + 1;
        }
        return sum;
    }

    [[nodiscard]] long part2(const std::vector<std::string> &lines) {
        auto width = lines[0].size();
        auto height = lines.size();
        const auto &heightmap = heightmap_from_file(lines);
        const auto &low_points = get_low_points(heightmap, width, height);
        auto visited = std::set<int>{};
        auto results = std::vector<std::set<int>>{};
        for (const auto &low_point: low_points) {
            auto basin = std::set<int>{};
            get_basin(low_point, heightmap, width, height, basin, visited);
            results.emplace_back(basin);
        }
        std::sort(results.begin(), results.end(),
                  [](const std::set<int> &a, const std::set<int> &b) {
                      return a.size() < b.size();
                  });
        return results[results.size() - 1].size() *
               results[results.size() - 2].size() *
               results[results.size() - 3].size();
    }
}

File: 2021/rust/src/day9.rs
use std::collections::HashSet;

fn create_heightmap(lines: &Vec<String>) -> Vec<i32> {
    lines
        .iter()
        .flat_map(|s| {
            s.chars()
                .map(|c| c.to_digit(10).unwrap() as i32)
                .collect::<Vec<_>>()
        })
        .collect()
}

fn index(x: i32, y: i32, width: i32) -> i32 {
    y * width + x
}

fn position(idx: i32, width: i32) -> (i32, i32) {
    (idx % width, idx / width)
}

fn get_neighbors(x: i32, y: i32, width: i32, height: i32) -> Vec<i32> {
    let mut neighbors = vec![];
    if x - 1 >= 0 {
        neighbors.push(index(x - 1, y, width));
    }
    if x + 1 < width {
        neighbors.push(index(x + 1, y, width));
    }
    if y - 1 >= 0 {
        neighbors.push(index(x, y - 1, width));
    }
    if y + 1 < height {
        neighbors.push(index(x, y + 1, width));
    }
    neighbors
}

fn get_low_points(heightmap: &Vec<i32>, width: i32, height: i32) -> HashSet<i32> {
    let mut low_points = HashSet::<i32>::new();
    for i in 0..heightmap.len() {
        let (x, y) = position(i as i32, width);
        let curr_height = heightmap[i];
        if curr_height == 9 {
            continue;
        }

        let neighbors = get_neighbors(x, y, width, height);
        if curr_height
            < neighbors
                .iter()
                .map(|n| heightmap[*n as usize])
                .min()
                .unwrap()
        {
            low_points.insert(i as i32);
        }
    }
    low_points
}

fn get_basin(
    idx: i32,
    heightmap: &Vec<i32>,
    width: i32,
    height: i32,
    basin: &mut HashSet<i32>,
    visited: &mut HashSet<i32>,
) {
    visited.insert(idx);
    let (x, y) = position(idx as i32, width);
    let curr_height = heightmap[idx as usize];
    if curr_height == 9 {
        return;
    }
    basin.insert(idx);
    let neighbors = get_neighbors(x, y, width, height);
    for neighbor in neighbors {
        if !visited.contains(&neighbor) && heightmap[neighbor as usize] != 9 {
            get_basin(neighbor, heightmap, width, height, basin, visited);
        }
    }
}

pub fn part1(lines: &Vec<String>) -> i32 {
    let width = lines[0].len();
    let height = lines.len();
    let heightmap = create_heightmap(lines);
    let low_points = get_low_points(&heightmap, width as i32, height as i32);
    let r = low_points
        .iter()
        .map(|lp| heightmap[*lp as usize] + 1)
        .sum::<i32>();
    r as i32
}

pub fn part2(lines: &Vec<String>) -> i32 {
    let width = lines[0].len();
    let height = lines.len();
    let heightmap = create_heightmap(lines);

    let mut results = vec![];
    let mut visited = HashSet::<i32>::new();
    for idx in get_low_points(&heightmap, width as i32, height as i32) {
        let mut basin = HashSet::<i32>::new();
        get_basin(
            idx,
            &heightmap,
            width as i32,
            height as i32,
            &mut basin,
            &mut visited,
        );
        results.push(basin);
    }
    results.sort_by(|a, b| a.len().partial_cmp(&b.len()).unwrap());
    results.reverse();
    results
        .iter()
        .take(3)
        .map(|a| a.len() as i32)
        .reduce(|a, b| a * b)
        .unwrap()
}
